### PR TITLE
core/vm: use uint64 in Memory for indices everywhere

### PR DESCRIPTION
**Pull request by lmittmann**
Consistently use `uint64` for indices in `Memory` and drop lots of type conversions from `uint64` to `int64`.

### CODE REVIEW

**Review by lightclient — COMMENTED** [review]
I'll let other chime in, but we're enacting a bit of a policy of not merging PRs like this that have minimal impacts on performance. While it does seem marginally cleaner to use `uint64` instead of `int64`, we could also just add a function `Int64` to the `uint256.Int` type and get roughly the same result.

Additionally, golang represents indexes and lengths as int. This would deviate from that paradigm.

**Comment by lmittmann** [comment]
This PR is primarily a cleanup, that I did not want to include into #30137 to keep its diff small.

> Additionally, golang represents indexes and lengths as int. This would deviate from that paradigm.

Yes, but Memory uses uint64 as indices almost everywhere, which is totally fine. There is just the two functions I changed in this PR that use int64 for no good reason resulting in cluttered code. This PR is the cleanup

**Review by rjl493456442 — APPROVED** [review]

**Review by lightclient — APPROVED** [review]

**Comment by lightclient** [comment]
cc @holiman

File: core/vm/memory.go
Comment (karalabe) [inline core/vm/memory.go:74]: This check seems to not be needed. We always expand the memory before trying to access it, so this check is useless (plus it was always wrong because it should have been offset + size). Might as well delete it now.

File: core/vm/memory.go
Comment (karalabe) [inline core/vm/memory.go:88]: This check seems to not be needed. We always expand the memory before trying to access it, so this check is useless (plus it was always wrong because it should have been offset + size). Might as well delete it now.

**Comment by karalabe** [comment]
We'll deploy it on a few benchmarkers to sanity check and will merge if nothing blows up :)

**Comment by rjl493456442** [comment]
Deployed it on a fully-synced node on https://my.papertrailapp.com/systems/13274397641/events
Deployed it for full sync from scratch on https://my.papertrailapp.com/systems/13274397401/events

Will let them run for a while

**Review by holiman — APPROVED** [review]
LGTM

**Pull request merged**
